### DISCUSSION TITLE
Static or Binary Scaler

### DISCUSSION THREAD

**Discussion by rumstead** [kedacore/keda#3374] [Q&A / Need Help] [answered]
The use case is to have the most naive scaler ever, on or off. Think of a primary and secondary where we want static replicas for our workloads on the primary and scaled-down replicas on the secondary. During a cluster upgrade, DR, etc we could make the secondary "active" and allow for the nodes and workloads to scale up. 

Has anyone implemented or knows if current scalers could fit the need? I have tried writing a custom scale. It scales down to 0 fine but it cannot get the workloads to stay at the maximum replicas.

**Comment by JorTurFer** [comment, marked as answer]
Taking into account that you need an external source to switch from one state to the other, why not providing the replica count from it?
I mean, you could use ie the Metrics API  and connect to a Rest API which manages the state between the options. Binary option could be exposing 0 or X

**Reply by rumstead** [reply]
> Taking into account that you need an external source to switch from one state to the other

Yup, that is exactly the plan (don't judge my playground code). Expose some endpoint to allow an external system to control when to swap. 
[Here](https://github.com/rumstead/keda-playground/blob/a623fbc188c2d5ba4a845bbd528218647919b459/main.go#L33) is a rest endpoint that [swaps](https://github.com/rumstead/keda-playground/blob/a623fbc188c2d5ba4a845bbd528218647919b459/pkg/scalers/static/static.go#L20). 

Can you tell the HPA/Keda to go to a specific replica count? I tried feeding it [dummy metrics](https://github.com/rumstead/keda-playground/blob/a623fbc188c2d5ba4a845bbd528218647919b459/pkg/scalers/static/static.go#L62) but that didn't scale the replicas to the max.

**Reply by JorTurFer** [reply]
I get your requirements. The point is that IMO there are already scalers and options that you can use to achieve that.
First of all, answering to your dummy scaler: There are 3 functions directly related with the autoscaling:
- `IsActive` is used by KEDA Operator to scale from/to 0
- `GetMetricsSpec` is used by KEDA Operator to get the values for creating the HPA
- `GetMetrics` is used for getting the value every time that it's needed to evaluate the autoscaling requirements.

In your case, [you should return 1 in `GetMetricsSpect`](https://github.com/rumstead/keda-playground/blob/a623fbc188c2d5ba4a845bbd528218647919b459/pkg/scalers/static/static.go#L62) and implement `GetMetrics` returning the static value that you want. With these changes, `IsActive` will decide to "turn on" the scaler and then the `GetMetrics` will return the value for the HPA controller to scale max.

Now, how I'd do it. I'd use a Scaler which upstream I can control, ie MetricsAPI or External Scaler. I will create a REST API or gRPC app the expose the value that I want on each time and I'd connect KEDA to it. You can see an scenario like that with [metrics-api e2e](https://github.com/kedacore/keda/blob/7c24d7bd2542e232508ff7cbafb3ec1e98734684/tests/scalers/metrics-api.test.ts#L69).
Basically, we use a REST API and create a SO that hits that API and doing post to the API, we can decide the amount of replicas that the HPA will create. If you don't want to create the REST API, [you could just annotate the SO with the pause annotation](https://keda.sh/docs/2.7/concepts/scaling-deployments/#pause-autoscaling) specifying the value that you want and KEDA will ensure that it's the instance count

**Reply by rumstead** [reply]
Appreciate you taking the time to explain and your feedback! I would much rather use a community-supported scaler than my own and that is exactly what I was looking for.

**Reply by JorTurFer** [reply]
Happy to help :)
Tell me if you have any other problem
